pkg/controller/util/endpoint/controller_utils.go

```diff
@@ -34,6 +36,76 @@ import (
 	"k8s.io/kubernetes/pkg/util/hash"
 )
 
+// ServiceSelectorCache is a cache of service selectors to avoid high CPU consumption caused by frequent calls to AsSelectorPreValidated (see #73527)
+type ServiceSelectorCache struct {
+	lock  sync.RWMutex
+	cache map[string]labels.Selector
+}
+
+// NewServiceSelectorCache init ServiceSelectorCache for both endpoint controller and endpointSlice controller.
+func NewServiceSelectorCache() *ServiceSelectorCache {
+	return &ServiceSelectorCache{
+		cache: map[string]labels.Selector{},
+	}
+}
+
+// Get return selector and existence in ServiceSelectorCache by key.
+func (sc *ServiceSelectorCache) Get(key string) (labels.Selector, bool) {
+	sc.lock.RLock()
+	selector, ok := sc.cache[key]
+	// fine-grained lock improves GetPodServiceMemberships performance(16.5%) than defer measured by BenchmarkGetPodServiceMemberships
+	sc.lock.RUnlock()
+	return selector, ok
+}
+
+// Update can update or add a selector in ServiceSelectorCache while service's selector changed.
+func (sc *ServiceSelectorCache) Update(key string, rawSelector map[string]string) labels.Selector {
+	sc.lock.Lock()
+	defer sc.lock.Unlock()
+	selector := labels.Set(rawSelector).AsSelectorPreValidated()
+	sc.cache[key] = selector
+	return selector
+}
+
+// Delete can delete selector which exist in ServiceSelectorCache.
+func (sc *ServiceSelectorCache) Delete(key string) {
+	sc.lock.Lock()
+	defer sc.lock.Unlock()
+	delete(sc.cache, key)
+}
+
+// GetPodServiceMemberships returns a set of Service keys for Services that have
+// a selector matching the given pod.
+func (sc *ServiceSelectorCache) GetPodServiceMemberships(serviceLister v1listers.ServiceLister, pod *v1.Pod) (sets.String, error) {
+	set := sets.String{}
+	services, err := serviceLister.Services(pod.Namespace).List(labels.Everything())
+	if err != nil {
+		return set, err
+	}
+
+	var selector labels.Selector
+	for _, service := range services {
+		if service.Spec.Selector == nil {
+			// if the service has a nil selector this means selectors match nothing, not everything.
+			continue
+		}
+		key, err := controller.KeyFunc(service)
+		if err != nil {
+			return nil, err
+		}
+		if v, ok := sc.Get(key); ok {
+			selector = v
+		} else {
+			selector = sc.Update(key, service.Spec.Selector)
+		}
+
+		if selector.Matches(labels.Set(pod.Labels)) {
+			set.Insert(key)
+		}
```

Comment:
@wojtek-t correct me if I'm wrong, but here the services are obtained from the Lister, it does not make any sense using the `controller.KeyFunc` that is an alias for `cache.DeletionHandlingMetaNamespaceKeyFunc`, right? you already get an object typed from the lister

Reply:
Not sure I understand the question here - yes, these are from lister, properly typed etc.
But IIUC we wanted to return their identifiers, and we're doing exactly that.
What alternative you have on your mind?

Reply:
I phrased it wrongly, sorry I wan to mean that to get the key of the service we can just do `key = fmt.Sprintf("%s/%s", service.Namespace, service.Name)` instead of adding a dependency on `controller`or `cache`

This is on the context of moving this code to staging, that less dependencies the better

Reply:
Yes - that should work.